modules/core/src/main/scala/Json.scala

```diff
@@ -19,28 +19,34 @@ trait JsonModule[R <: Realisation] extends SchemaModule[R] {
     implicit primNT: R.Prim ~> Encoder,
     fieldLabel: R.ProductTermId <~< String,
     branchLabel: R.SumTermId <~< String
-  ): HAlgebra[Schema[R.Prim, R.SumTermId, R.ProductTermId, ?[_], ?], Encoder] =
-    new (Schema[R.Prim, R.SumTermId, R.ProductTermId, Encoder, ?] ~> Encoder) {
-
-      val encloseInBraces         = (s: String) => s"{$s}"
-      def makeField(name: String) = (s: String) => s""""$name":$s"""
-
-      def apply[A](schema: Schema[R.Prim, R.SumTermId, R.ProductTermId, Encoder, A]): Encoder[A] =
-        schema match {
-
-          case PrimSchema(prim) => primNT(prim)
-          case :*:(left, right) => (a => left(a._1) + "," + right(a._2))
-          case :+:(left, right) => (a => a.fold(left, right))
-          case i: IsoSchema[R.Prim, R.SumTermId, R.ProductTermId, Encoder, _, A] =>
-            i.base.compose(i.iso.reverseGet)
-          case r: Record[R.Prim, R.SumTermId, R.ProductTermId, Encoder, A, _] =>
-            encloseInBraces.compose(r.fields).compose(r.iso.reverseGet)
-          case SeqSchema(element)    => (a => a.map(element).mkString("[", ",", "]"))
-          case ProductTerm(id, base) => makeField(fieldLabel(id)).compose(base)
-          case u: Union[R.Prim, R.SumTermId, R.ProductTermId, Encoder, A, _] =>
-            encloseInBraces.compose(u.choices).compose(u.iso.reverseGet)
-          case SumTerm(id, base) => makeField(branchLabel(id)).compose(base)
-          case One()             => (_ => "null")
-        }
+  ): RInterpreter[Encoder] =
+    new Interpreter[R.Prim, R.SumTermId, R.ProductTermId, Encoder] {
+
+      val alg: HAlgebra[RSchema, Encoder] = new (RSchema[Encoder, ?] ~> Encoder) {
+
+        val encloseInBraces         = (s: String) => s"{$s}"
+        def makeField(name: String) = (s: String) => s""""$name":$s"""
+
+        def apply[A](schema: RSchema[Encoder, A]): Encoder[A] =
+          schema match {
+
+            case PrimSchema(prim) => primNT(prim)
+            case :*:(left, right) => (a => left(a._1) + "," + right(a._2))
+            case :+:(left, right) => (a => a.fold(left, right))
+            case i: IsoSchema[R.Prim, R.SumTermId, R.ProductTermId, Encoder, _, A] =>
+              i.base.compose(i.iso.reverseGet)
+            case r: Record[R.Prim, R.SumTermId, R.ProductTermId, Encoder, A, _] =>
+              encloseInBraces.compose(r.fields).compose(r.iso.reverseGet)
+            case SeqSchema(element)    => (a => a.map(element).mkString("[", ",", "]"))
+            case ProductTerm(id, base) => makeField(fieldLabel(id)).compose(base)
+            case u: Union[R.Prim, R.SumTermId, R.ProductTermId, Encoder, A, _] =>
+              encloseInBraces.compose(u.choices).compose(u.iso.reverseGet)
+            case SumTerm(id, base) => makeField(branchLabel(id)).compose(base)
+            case One()             => (_ => "null")
+          }
```

Comment:
This is a bit unrelated to this PR but what is the difference between `:*:`, `Record` and `ProductTerm`?

Reply:
`:*:(a, b)` represents a `(A, B)`, `ProductTerm(id, a)` represents a "field" with name `id` and schema `a` (it doesn't make much sense outside of a record).
A `Record` contains a schema that is guarantied to be composed of a "tree of `:*:`" whose all left children (as well as the right child of the deepest `:*:`) are `ProductTerm`s.

Reply:
> A `Record` contains a schema that is guarantied to be composed of a "tree of `:*:`" whose all left children (as well as the right child of the deepest `:*:`) are `ProductTerm`s.

Could we see this guarantee in the types? Currently the type of a record’s `fields` is `F[_]`.

Reply:
I don't think it's doable. We need `fields` to be a `F[_]` to use recursion schemes (during a `cataNT` for example, `F` is bound to the "target functor" the schema is folded to).